File: src/backend/tracking-service/internal/models/location.go
package models

import (
	// time for handling timestamps and ensuring temporal accuracy in location records (go1.21)
	"time"
	// json package for efficient JSON serialization and deserialization of location data with error handling (go1.21)
	"encoding/json"
	// uuid package for generating UUID v4 (v1.3.0)
	"github.com/google/uuid"
)

// MinLatitude represents the minimum valid latitude coordinate.
const MinLatitude float64 = -90.0

// MaxLatitude represents the maximum valid latitude coordinate.
const MaxLatitude float64 = 90.0

// MinLongitude represents the minimum valid longitude coordinate.
const MinLongitude float64 = -180.0

// MaxLongitude represents the maximum valid longitude coordinate.
const MaxLongitude float64 = 180.0

// DefaultAccuracy defines the default GPS accuracy in meters.
const DefaultAccuracy float64 = 10.0

// MaxAccuracy defines the maximum acceptable GPS accuracy in meters.
const MaxAccuracy float64 = 100.0

// Location represents a single GPS location point with coordinates, timestamp,
// and accuracy metrics. It includes comprehensive validation, serialization,
// and data integrity checks for real-time tracking scenarios.
type Location struct {
	// ID is a unique identifier for the location point (UUID v4).
	ID string `json:"id"`

	// WalkID links this location to a specific dog walk session.
	WalkID string `json:"walkId"`

	// Latitude represents the latitude coordinate associated with this location.
	Latitude float64 `json:"latitude"`

	// Longitude represents the longitude coordinate associated with this location.
	Longitude float64 `json:"longitude"`

	// Accuracy indicates the positional accuracy in meters.
	Accuracy float64 `json:"accuracy"`

	// Altitude represents the height above sea level in meters.
	Altitude float64 `json:"altitude"`

	// Timestamp captures the exact time this location was recorded, in UTC.
	Timestamp time.Time `json:"timestamp"`

	// IsValid indicates whether the current location data has passed validation.
	IsValid bool `json:"isValid"`
}

// NewLocation creates a new Location instance with comprehensive validation
// and initialization. The function returns an error if any validation step fails.
func NewLocation(
	walkID string,
	latitude float64,
	longitude float64,
	accuracy float64,
	altitude float64,
) (Location, error) {

	var loc Location

	// Generate unique ID using UUID v4
	newID := uuid.NewString()
	loc.ID = newID

	// Assign WalkID, ensuring it is not empty
	loc.WalkID = walkID

	// Set latitude and longitude
	loc.Latitude = latitude
	loc.Longitude = longitude

	// Choose accuracy: if zero is provided, default to DefaultAccuracy
	if accuracy == 0 {
		loc.Accuracy = DefaultAccuracy
	} else {
		loc.Accuracy = accuracy
	}

	// Set altitude
	loc.Altitude = altitude

	// Set timestamp to current UTC time
	loc.Timestamp = time.Now().UTC()

	// Perform a full validation on the location
	if err := loc.Validate(); err != nil {
		return loc, err
	}

	// If all validation checks pass, IsValid will be true
	return loc, nil
}

// Validate performs comprehensive checks on the Location fields:
//  1. ID must be a valid UUID.
//  2. WalkID cannot be empty.
//  3. Latitude must be within [-90.0, 90.0].
//  4. Longitude must be within [-180.0, 180.0].
//  5. Accuracy must be within [0.0, MaxAccuracy].
//  6. Timestamp must be non-zero and not significantly in the future.
func (l *Location) Validate() error {
	// Verify ID is valid UUID
	if _, parseErr := uuid.Parse(l.ID); parseErr != nil {
		l.IsValid = false
		return parseErr
	}

	// Check that WalkID is not empty
	if l.WalkID == "" {
		l.IsValid = false
		return ErrInvalidWalkID("WalkID cannot be empty")
	}

	// Check latitude bounds
	if l.Latitude < MinLatitude || l.Latitude > MaxLatitude {
		l.IsValid = false
		return ErrOutOfRange("Latitude is out of valid range")
	}

	// Check longitude bounds
	if l.Longitude < MinLongitude || l.Longitude > MaxLongitude {
		l.IsValid = false
		return ErrOutOfRange("Longitude is out of valid range")
	}

	// Verify accuracy range
	if l.Accuracy < 0.0 || l.Accuracy > MaxAccuracy {
		l.IsValid = false
		return ErrOutOfRange("Accuracy is out of valid range")
	}

	// Ensure timestamp is not the zero value
	if l.Timestamp.IsZero() {
		l.IsValid = false
		return ErrInvalidTimestamp("Timestamp cannot be zero")
	}

	// Disallow timestamps significantly in the future
	nowUTC := time.Now().UTC()
	if l.Timestamp.After(nowUTC.Add(1 * time.Minute)) {
		l.IsValid = false
		return ErrInvalidTimestamp("Timestamp is set too far in the future")
	}

	// If all checks pass
	l.IsValid = true
	return nil
}

// ToJSON serializes the Location data into its JSON representation. If any
// validation step fails, an error will be returned instead of JSON data.
func (l *Location) ToJSON() ([]byte, error) {
	// Validate location data before serialization
	if err := l.Validate(); err != nil {
		return nil, err
	}

	// Marshal Location struct to JSON
	jsonData, err := json.Marshal(l)
	if err != nil {
		return nil, err
	}
	return jsonData, nil
}

// FromJSON creates a new Location instance from the provided JSON data,
// performs complete validation, and returns an error if any step fails.
func FromJSON(data []byte) (Location, error) {
	var loc Location

	// Unmarshal JSON data into a temporary Location struct
	if err := json.Unmarshal(data, &loc); err != nil {
		return loc, err
	}

	// Validate the newly populated fields
	if err := loc.Validate(); err != nil {
		return loc, err
	}

	return loc, nil
}

// ErrInvalidWalkID is returned when the provided WalkID is empty or invalid.
type ErrInvalidWalkID string

func (e ErrInvalidWalkID) Error() string {
	return string(e)
}

// ErrOutOfRange is returned when a numeric field such as latitude, longitude,
// or accuracy is outside of acceptable bounds.
type ErrOutOfRange string

func (e ErrOutOfRange) Error() string {
	return string(e)
}

// ErrInvalidTimestamp is returned when the timestamp field is invalid.
type ErrInvalidTimestamp string

func (e ErrInvalidTimestamp) Error() string {
	return string(e)
}

File: src/android/app/src/main/java/com/dogwalking/app/domain/models/Location.kt
package com.dogwalking.app.domain.models

import android.os.Parcelable // android.os version default – Facilitates inter-component data serialization in Android
import kotlinx.parcelize.Parcelize // kotlinx.parcelize version 1.9.0 – Provides the @Parcelize annotation for automatic Parcelable generation
import com.google.gson.annotations.SerializedName // com.google.gson.annotations version 2.9.0 – Enables field-level JSON serialization naming
import com.google.gson.Gson // com.google.gson version 2.9.0 – Core JSON processing library
import com.google.gson.GsonBuilder
import com.google.gson.TypeAdapter
import com.google.gson.stream.JsonReader
import com.google.gson.stream.JsonWriter
import java.util.Locale

/**
 * Data class representing a GPS location point with essential tracking information for dog walks.
 * This model implements Parcelable for efficient data sharing between Android components.
 * It includes comprehensive validation logic to ensure location data integrity, conforming
 * to all business requirements for real-time GPS tracking and walk monitoring.
 */
@Parcelize
data class Location(
    /**
     * Unique identifier for this GPS location entry.
     * Must be non-blank and used to correlate with database records.
     */
    @SerializedName("id")
    val id: String,

    /**
     * Unique identifier referencing the dog walk session this location belongs to.
     * Must be non-blank and used to associate location data with specific active or past walks.
     */
    @SerializedName("walkId")
    val walkId: String,

    /**
     * Latitude in degrees. Valid range: -90.0 to 90.0.
     * Values outside this range will fail validation.
     */
    @SerializedName("latitude")
    val latitude: Double,

    /**
     * Longitude in degrees. Valid range: -180.0 to 180.0.
     * Values outside this range will fail validation.
     */
    @SerializedName("longitude")
    val longitude: Double,

    /**
     * Accuracy of the location measurement in meters.
     * Must be a non-negative value less than 100 meters to be considered valid.
     */
    @SerializedName("accuracy")
    val accuracy: Float,

    /**
     * Speed of the user/device in meters per second.
     * Must be a non-negative value less than 30 m/s to be considered valid.
     */
    @SerializedName("speed")
    val speed: Float,

    /**
     * Timestamp (in milliseconds since epoch) indicating when this location was recorded.
     * Invalid if it is more than 5 minutes old or is set in the future.
     */
    @SerializedName("timestamp")
    val timestamp: Long

) : Parcelable {

    /**
     * Validates if the location data is within acceptable bounds and meets
     * all integrity requirements for dog walking tracking scenarios.
     *
     * @return True if all location fields are valid; false otherwise.
     */
    fun isValid(): Boolean {
        // Check valid latitude: must be between -90 and 90 degrees.
        if (latitude < -90.0 || latitude > 90.0) {
            return false
        }
        // Check valid longitude: must be between -180 and 180 degrees.
        if (longitude < -180.0 || longitude > 180.0) {
            return false
        }
        // Check accuracy: must be >= 0 and less than 100 meters.
        if (accuracy < 0.0f || accuracy >= 100.0f) {
            return false
        }
        // Check speed: must be >= 0 and less than 30 m/s (approx 108 km/h, well above normal walking).
        if (speed < 0.0f || speed >= 30.0f) {
            return false
        }

        // Ensure timestamp is not more than 5 minutes in the past and not set in the future.
        val currentTime = System.currentTimeMillis()
        val fiveMinutesInMs = 5 * 60 * 1000
        if (timestamp > currentTime) {
            return false
        }
        if (timestamp < (currentTime - fiveMinutesInMs)) {
            return false
        }

        // Validate that both 'id' and 'walkId' are not blank.
        if (id.isBlank() || walkId.isBlank()) {
            return false
        }

        // Return true only if all the above checks pass.
        return true
    }

    /**
     * Converts this location data into a JSON string.
     * Only produces a valid JSON representation if the data passes all validation checks.
     * Provides precise number formatting via custom TypeAdapters for Double and Float fields.
     *
     * @return JSON string representation of the location data. Returns "{}" if invalid.
     */
    fun toJson(): String {
        // If the location is invalid, return an empty JSON object.
        if (!isValid()) {
            return "{}"
        }

        // Build a Gson instance with custom adapters for controlling numeric output formatting.
        val gsonWithCustomAdapters: Gson = GsonBuilder()
            .registerTypeAdapter(Double::class.java, object : TypeAdapter<Double>() {
                /**
                 * Writes a Double value with a fixed 6-decimal precision to ensure uniform output.
                 */
                override fun write(out: JsonWriter, value: Double?) {
                    if (value == null) {
                        out.nullValue()
                    } else {
                        out.value(String.format(Locale.US, "%.6f", value))
                    }
                }

                /**
                 * Reads a Double value from the JSON stream.
                 * This implementation simply parses the numeric string.
                 */
                override fun read(`in`: JsonReader): Double {
                    return try {
                        `in`.nextString().toDouble()
                    } catch (e: NumberFormatException) {
                        0.0
                    }
                }
            })
            .registerTypeAdapter(Float::class.java, object : TypeAdapter<Float>() {
                /**
                 * Writes a Float value with a fixed 2-decimal precision to ensure uniform output.
                 */
                override fun write(out: JsonWriter, value: Float?) {
                    if (value == null) {
                        out.nullValue()
                    } else {
                        out.value(String.format(Locale.US, "%.2f", value))
                    }
                }

                /**
                 * Reads a Float value from the JSON stream.
                 * This implementation handles possible NumberFormatExceptions gracefully.
                 */
                override fun read(`in`: JsonReader): Float {
                    return try {
                        `in`.nextString().toFloat()
                    } catch (e: NumberFormatException) {
                        0.0f
                    }
                }
            })
            .disableHtmlEscaping() // Ensures characters like '<', '>' are properly preserved.
            .create()

        // Return the serialized JSON representation using the custom GSON instance.
        return gsonWithCustomAdapters.toJson(this)
    }
}